src/ultk/language/semantics.py

```diff
@@ -43,6 +43,12 @@ def __str__(self) -> str:
     def __lt__(self, other):
         return self.name < other.name
 
+    def __eq__(self, other) -> bool:
+        return self.name == other.name and self.__dict__ == other.__dict__
+
+    def __hash__(self) -> int:
+        return hash((self.name, tuple(self.__dict__)))
```

Comment:
Is it possible that the same object could be hashed differently since `tuple` is ordered but `dict` is not by default?

Reply:
Dict should be ordered by insertion order in Python 3.7